### PR TITLE
Parse XML sitemaps via SAX for option to parse partial/broken documents

### PR DESCRIPTION
**Pull request by matt-deboer**
Adds an additional optional constructor arg (false by default) which enables parsing of partial/broken xml-based sitemap documents.

### CODE REVIEW

**Comment by jnioche** [comment]
Thanks @matt-deboer, looks interesting. Do you expect any performance gains compared to the current DOM-based parser?
Could you please apply the formattting as per [https://github.com/crawler-commons/crawler-commons/wiki/Code-Formatting] and squash your commits? This will make it easier to review this PR.
Thanks

**Comment by matt-deboer** [comment]
Sure, done.

In general, performance gains are expected, especially for very large sitemaps (haven't measured that aspect yet; I'll reply back with some quick measurements)--the real driver was to simplify the case where a broken sitemap was delivered (I have a particular focused crawl in mind); with DOM, I had to manually patch the contents and re-process them. Yeah, one could argue that broken sitemaps shouldn't exist, but not everyone you want/have to crawl is up to speed on their SEO : )

There could be more gains with the parsing of really large sitemaps if an additional public **parseSiteMap** method were added to take in an InputStream instead of byte[] -- allowing the caller to avoid pulling the whole contents into memory at once.

What do you think?

**Comment by Chaiavi** [comment]
Is it needed optimization ?

As the limit for a sitemap is 10mb

**Comment by matt-deboer** [comment]
This is not primarily a perf optimization--it is a "robustness" improvement
to allow parsing of broken site maps. Any perf improvement is an added
bonus...

**Comment by kkrugler** [comment]
Hi Matt,

I think Avi was asking specifically for any additional details about:

> > > There could be more gains with the parsing of really large sitemaps if an
> > > additional public *parseSiteMap* method were added to take in an
> > > InputStream instead of byte[] -- allowing the caller to avoid pulling the
> > > whole contents into memory at once.

e.g. have you seen issues with having to load the sitemap into memory first.

But in any case, seems very useful to handle broken site maps, since as you note it's common to get busted files in the real world.

Also, if there were no regressions w/using a SAX-based approach, then what would be reasons for sticking with the existing DOM approach?

Thanks,

-- Ken


> From: Matt DeBoer
> Sent: May 28, 2015 9:44:18am PDT
> To: crawler-commons/crawler-commons
> Subject: Re: [crawler-commons] Parse XML sitemaps via SAX for option to parse partial/broken documents (#80)
> 
> This is not primarily a perf optimization--it is a "robustness" improvement
> to allow parsing of broken site maps. Any perf improvement is an added
> bonus...
> 
> On Thursday, May 28, 2015, Avi Hayun <notifications@github.com> wrote:
> 
> > Is it needed optimization ?
> >
> > As the limit for a sitemap is 10mb
> >
> > On Thu, May 28, 2015 at 3:00 AM, Matt DeBoer <notifications@github.com
> > <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> > wrote:
> >
> > > Sure, done.
> > >
> > > In general, performance gains are expected, especially for very large
> > > sitemaps (haven't measured that aspect yet; I'll reply back with some
> > quick
> > > measurements)--the real driver was to simplify the case where a broken
> > > sitemap was delivered (I have a particular focused crawl in mind); with
> > > DOM, I had to manually patch the contents and re-process them. Yeah, one
> > > could argue that broken sitemaps shouldn't exist, but not everyone you
> > > want/have to crawl is up to speed on their SEO : )
> > >
> > > There could be more gains with the parsing of really large sitemaps if an
> > > additional public *parseSiteMap* method were added to take in an
> > > InputStream instead of byte[] -- allowing the caller to avoid pulling the
> > > whole contents into memory at once.
> > >
> > > What do you think?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> > https://github.com/crawler-commons/crawler-commons/pull/80#issuecomment-106115526
> > >
> > > .
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/crawler-commons/crawler-commons/pull/80#issuecomment-106218177>
> > .
> >
> —
> Reply to this email directly or view it on GitHub.
> 

--------------------------
Ken Krugler
+1 530-210-6378
http://www.scaleunlimited.com
custom big data solutions & training
Hadoop, Cascading, Cassandra & Solr





--------------------------
Ken Krugler
+1 530-210-6378
http://www.scaleunlimited.com
custom big data solutions & training
Hadoop, Cascading, Cassandra & Solr

**Comment by jnioche** [comment]
Would this solve #79?

**Comment by matt-deboer** [comment]
Sorry for the extended delay; just got back to this tonight.

@Chaiavi, @kkrugler: 
My original guesses about potential gains and "large" sitemaps proved to be wrong. After some testing with 10MB sitemaps, the performance is roughly the same (by roughly, I mean averaged about the same after 10 runs of each on my macbook). In fact, the initial implementation using a Stack was actually worse, so I swapped it for a LinkedList. My earlier comment about adding an additional *parseSiteMap* method can be disregarded.
To sum up, this PR is all about parsing of potentially broken site maps. 

@jnioche: 
I added a test case that included your example referenced in #79 (from #75), and your example is parsed successfully in this new implementation.

**Comment by lewismc** [comment]
Apart from this patch offering conflicts with master, is there a compelling reason not to update and merge folks?

**Comment by jnioche** [comment]
Would it make sense to have that as well as the DOM-based parser for some time? Could for instance have a separate package for this

**Comment by kkrugler** [comment]
Hi @jnioche - thanks for following up on this (old) issue...
